vassal-app/src/main/java/VASSAL/counters/Deck.java

```diff
@@ -1769,9 +1913,51 @@ public String getSelectSortProperty() {
     return selectSortProperty;
   }
 
-  protected void repaintMap() {
+  public void repaintMap() {
     if (map != null) {
       map.repaint();
     }
   }
+
+
+  /**
+   * Sort the Deck. Called from DeckSortKeyCommand
+   * @param sortParameters A list of parameters describing the sort
+   * @param reportFormat A report format to generate when sort performed
+   * @return Command encapsulating sort
+   */
+  public Command sort(List<SortParameter> sortParameters, FormattedString reportFormat, String menuCommand) {
+    Command c = new NullCommand();
+
+    // Do not even try unless there is at least one SortParameter with a non-null property name
+    if (sortParameters == null || sortParameters.isEmpty() || sortParameters.get(0).getSortProperty().isEmpty()) {
+      return c;
+    }
+
+    // Create a sortable list containing all the pieces
+    final SortablePiece[] pieces = new SortablePiece[getPieceCount()];
+    for (int i = 0; i < pieces.length; ++i) {
+      pieces[pieces.length - i - 1] = new SortablePiece(sortParameters, getPieceAt(i));
+    }
+
+    // Sort it!
+    Arrays.sort(pieces);
+
+    // Convert back to a list of GamePieces
+    final List<GamePiece> pieceList = new ArrayList<>();
+    for (final SortablePiece sp : pieces) {
+      pieceList.add(sp.getPiece());
+    }
+
```

Comment:
What's the reason for going back and forth between arrays and lists here?

Reply:
Yeah, fair cop. I was lazy and copied the existing sort code used in other part of Deck, which is based on the Array implementation of the contents of a Stack,  I need to end up with a List of GamePiece, so much better to convert the contents array to a List and refactor my SortablePiece class into a Comparator<GamePiece> to pass to Collections.Sort. I will work on this.

Reply:
Re-coded using Comparator and single conversion from Array to List